File: app/src/main/java/com/lemonade/dima/di/OkHttpProvider.java
package com.lemonade.dima.di;

import android.support.annotation.NonNull;

import com.lemonade.dima.api.MockingInterceptor;

import okhttp3.OkHttpClient;

public final class OkHttpProvider {
    private static volatile OkHttpClient sClient;

    @NonNull
    public static OkHttpClient provideClient() {
        OkHttpClient client = sClient;
        if (client == null) {
            synchronized (ApiFactory.class) {
                client = sClient;
                if (client == null) {
                    client = sClient = buildClient();
                }
            }
        }
        return client;
    }

    public static void recreate() {
        sClient = null;
        sClient = buildClient();
    }

    @NonNull
    private static OkHttpClient buildClient() {
        return new OkHttpClient.Builder()
                .addInterceptor(MockingInterceptor.create())
                .build();
    }
}

File: app/src/main/java/com/k4dima/chat/features/messages/api/OkHttpProvider.kt
package com.k4dima.chat.features.messages.api

import com.k4dima.chat.core.api.MockingInterceptor
import com.k4dima.chat.features.messages.di.MessagesScope
import dagger.Binds
import dagger.Module
import dagger.Provides
import okhttp3.Interceptor
import okhttp3.OkHttpClient

@Module
internal abstract class OkHttpProvider {
    @MessagesScope
    @Binds
    internal abstract fun create(mockingInterceptor: MockingInterceptor): Interceptor

    @Module
    companion object {
        @JvmStatic
        @MessagesScope
        @Provides
        fun buildClient(interceptor: Interceptor): OkHttpClient {
            return OkHttpClient.Builder()
                    .addInterceptor(interceptor)
                    .build()
        }
    }
}